Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,189 @@ lemma iff_flat_and_lTensor_reflects_triviality :
+    -- suppose `x ∈ E ⊗ M`, we aim to show `x = 0`
+    ext x
+    simp only [Submodule.mem_top, Submodule.mem_bot, true_iff]
+    have mem : x ∈ (⊤ : Submodule R _) := ⟨⟩
+    rw [← TensorProduct.span_tmul_eq_top, mem_span_set] at mem
+    obtain ⟨c, hc, rfl⟩ := mem
+    choose b a hy using hc
+    let r :  ⦃a : E ⊗[R] M⦄ → a ∈ ↑c.support → R := fun a ha =>
+      Submodule.mem_span_singleton.1 (b ha).2 |>.choose
+    have hr : ∀ ⦃i : E ⊗[R] M⦄ (hi : i ∈ c.support), b hi =
+        r hi • ⟨l23 (l12 n1), Submodule.mem_span_singleton_self _⟩ := by
+      intro i hi
+      ext
+      exact Submodule.mem_span_singleton.1 (b hi).2 |>.choose_spec.symm
+    -- Since `x ∈ E ⊗ M`, since `M` is flat and `E -> N1` is injective, we only need to check the
+    -- equality in `N1 ⊗ M`. we write `x = ∑ μᵢ • (l23 (l12 n1)) ⊗ mᵢ = ∑ μᵢ • 0 = 0`
+    -- (remember `E = span {l23 (l12 n1)}` and `eq1`)
+    refine Finset.sum_eq_zero fun i hi => show c i • i = 0 from
+      (Module.Flat.rTensor_preserves_injective_linearMap (M := M) E.subtype <|
+              Submodule.injective_subtype E) ?_
+    rw [← hy hi, hr hi, smul_tmul, map_smul, LinearMap.rTensor_tmul, Submodule.subtype_apply, eq1,
+      smul_zero, map_zero]
+  have : Subsingleton (E ⊗[R] M) := subsingleton_iff_forall_eq 0 |>.2 fun x =>
+    show x ∈ (⊥ : Submodule R _) from eq0 ▸ ⟨⟩
+
+  -- but `E ⊗ M = 0` implies `E = 0` because `M` is faithfully flat and this is a contradiction.
+  exact not_subsingleton_iff_nontrivial.2 inferInstance <| fl.rTensor_reflects_triviality R M E
+
+lemma rTensor_reflects_exact [fl : FaithfullyFlat R M]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+    (ex : Function.Exact (l12.rTensor M) (l23.rTensor M)) :
+    Function.Exact l12 l23 := LinearMap.exact_iff.2 <| by
+  have complex : LinearMap.range l12 ≤ LinearMap.ker l23 := range_le_ker_of_exact_rTensor R M _ _ ex
+  -- by previous lemma, we need only to show `ker l23 ≤ range l12`.
+  refine le_antisymm ?_ complex
+  -- We consider the cohomology of `N1 -> N2 -> N3`, i.e. let `H := ker l23 ⧸ range l12`. It is
+  -- enough to show `H = 0` because `H = 0` implies `x ∈ range l12`.
+  let H := LinearMap.ker l23 ⧸ LinearMap.range (Submodule.inclusion complex)
+  suffices triv_coh : Subsingleton H by
+    rintro x (hx : l23 x = 0)
+    have eq0 : (Submodule.mkQ _ ⟨x, hx⟩ : H) = 0 := triv_coh.elim _ _
+    obtain ⟨⟨y, hy⟩, eq0⟩ := Submodule.Quotient.mk_eq_zero _ |>.1 eq0
+    simp only [Subtype.ext_iff, Submodule.coe_inclusion] at eq0
+    exact eq0 ▸ hy
+
+  -- Since `M` is faithfully flat, we need only to show that `H ⊗ M` is trivial.
+  suffices Subsingleton (H ⊗[R] M) from rTensor_reflects_triviality R M H
+  let e : H ⊗[R] M ≃ₗ[R] _ := TensorProduct.quotientTensorEquiv _ _
+  rw [e.toEquiv.subsingleton_congr, Submodule.subsingleton_quotient_iff_eq_top, eq_top_iff]
+  -- note that `H ⊗ M ≅ (ker l23 ⊗ M) ⧸ (range l12 ⊗ M)`, so we only need to show that
+  -- `ker l23 ⊗ M ≤ range l12 ⊗ M` as submodules of `ker l23 ⊗ M`.
```

Comment:
Does this make sense? Do you mean submodules of something else? N2 tensor M? And again my feeling is that you should be proving equality here. It just looks weird (to me) to keep talking about having to show that B <= A instead of B = A when you have B/A in play meaning that A<=B.

Reply:
I have changed the strategy in this proof, please check if it makes more sense now.